src/tracing.ts

```diff
@@ -0,0 +1,173 @@
+  private readonly traceAgent: TraceAgent;
+  /** A logger. */
+  private readonly logger: common.Logger;
+  /** The configuration object for this instance. */
+  private readonly config: Forceable<NormalizedConfig>;
+
+  /**
+   * Constructs a new Tracing instance.
+   * @param config The configuration for this instance.
+   */
+  constructor(config: NormalizedConfig, traceAgent: TraceAgent) {
+    this.config = config;
+    this.traceAgent = traceAgent;
+    let logLevel = config.enabled ? config.logLevel : 0;
+    // Clamp the logger level.
+    if (logLevel < 0) {
+      logLevel = 0;
+    } else if (logLevel >= common.logger.LEVELS.length) {
+      logLevel = common.logger.LEVELS.length - 1;
+    }
+    this.logger = common.logger({
+      level: common.logger.LEVELS[logLevel],
+      tag: '@google-cloud/trace-agent'
+    });
+  }
+
+
+
+  /**
+   * Logs an error message detailing the list of modules that were loaded before
+   * the Trace Agent. Loading these modules before the Trace Agent may prevent
+   * us from monkeypatching those modules for automatic tracing.
+   * @param filesLoadedBeforeTrace The list of files that were loaded using
+   * require() before the Stackdriver Trace Agent was required.
+   */
+  logModulesLoadedBeforeTrace(filesLoadedBeforeTrace: string[]) {
+    const modulesLoadedBeforeTrace: string[] = [];
+    const traceModuleName = path.join('@google-cloud', 'trace-agent');
+    for (let i = 0; i < filesLoadedBeforeTrace.length; i++) {
+      const moduleName = packageNameFromPath(filesLoadedBeforeTrace[i]);
+      if (moduleName && moduleName !== traceModuleName &&
+          modulesLoadedBeforeTrace.indexOf(moduleName) === -1) {
+        modulesLoadedBeforeTrace.push(moduleName);
+      }
+    }
+    if (modulesLoadedBeforeTrace.length > 0) {
+      this.logger.error(
+          'TraceAgent#start: Tracing might not work as the following modules',
+          'were loaded before the trace agent was initialized:',
+          `[${modulesLoadedBeforeTrace.sort().join(', ')}]`);
+    }
+  }
+
+  /**
+   * Enables automatic tracing support and the custom span API.
+   */
+  enable(): void {
+    if (this.traceAgent.isActive()) {
+      // For unit tests only.
+      // Undoes initialization that occurred last time this function was called.
+      this.disable();
+    }
+
+    if (!this.config.enabled) {
+      return;
+    }
+
+    try {
+      // Initialize context propagation mechanism.
+      const clsConfig: Forceable<TraceCLSConfig> = {
+        mechanism: this.config.clsMechanism as TraceCLSMechanism,
+        [FORCE_NEW]: this.config[FORCE_NEW]
+      };
+      cls.create(clsConfig, this.logger).enable();
+
+      traceWriter.create(this.config, this.logger).initialize((err) => {
+        if (err) {
+          this.disable();
+        }
+      });
+
+      this.traceAgent.enable(this.config, this.logger);
+
+      pluginLoader.create(this.config, this.logger).activate();
```

Comment:
These two statements occur after we have called `TraceWriter#initialize` – however they are independent of success/failure of `traceWriter#initialize` as they are not in the callback block. From a clarity point of view, can you make one of the following statements true:

1. If these are not dependent on the outcome of `traceWriter#initialize`, move the async operation (`traceWriter#initialize`) the last operation, and call `traceAgent.enable` and `pluginLoader.create` first.
1. If these are dependent, move them into the callback for `create`.

As written, it is not obvious whether it is an oversight – should they have been inside the callback block?

Reply:
In response to the next comment I moved the initialization of the trace agent and plugin loader outside of the try-catch, so this probably doesn't apply anymore.